### PR TITLE
Initial support for Conduit

### PR DESCRIPTION
**Pull request by cvwright**
I used the `matrix-dendrite` role as a template to create a new role for Conduit.

I skipped all of the TURN configuration and similar stuff in `group_vars` because I don't really understand how it works.

To try the new role, simply set your host's `homeserver_implementation` to `conduit`.

### CODE REVIEW

**Comment by kubo6472** [comment]
feel free to take anything usable out of my PR, it is in separate branch, so should be easy to cherrypick

**Comment by kubo6472** [comment]
seems like you're already a lot further than I was.

File: roles/matrix-conduit/templates/conduit/conduit.toml.j2 [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+# your web server (e.g. apache or nginx), so all requests to /_matrix on port
+# 443 and 8448 will be forwarded to the Conduit instance running on this port
+# Docker users: Don't change this, you'll need to map an external port to this.
+port = {{ matrix_conduit_port_number }}
+
+# Max size for uploads
+max_request_size = {{ matrix_conduit_max_request_size }}
+
+# Enables registration. If set to false, no users can register on this server.
+allow_registration = {{ matrix_conduit_allow_registration | to_json | replace("\"", "") }}
```
Comment (spantaleev) [inline roles/matrix-conduit/templates/conduit/conduit.toml.j2:38]: `matrix_conduit_allow_registration` is a bool, isn't it. If so, ` | to_json` should do the right thing and not quote it as a string.

I suppose it's a string in your case (somehow) and your ` | replace` call tries to remedy that. I don't see why it's necessary.

----

The same applies to `allow_federation` and `enable_lightning_bolt` below.

File: roles/matrix-conduit/templates/conduit/conduit.toml.j2 [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+allow_federation = {{ matrix_conduit_allow_federation | to_json | replace("\"", "") }}
+
+# Enable the display name lightning bolt on registration.
+enable_lightning_bolt = {{ matrix_conduit_enable_lightning_bolt | to_json | replace("\"", "") }}
+
+trusted_servers = {{ matrix_conduit_trusted_servers }}
```
Comment (spantaleev) [inline roles/matrix-conduit/templates/conduit/conduit.toml.j2:45]:
```suggestion
trusted_servers = {{ matrix_conduit_trusted_servers | to_json }}
```

This is a list in `defaults/main.yml` and in [conduit-example.toml](https://gitlab.com/famedly/conduit/-/blob/0b926c2a31deff57a3526dd75d8c08775b02241a/conduit-example.toml#L46), so I suppose we can just use ` | to_json` here and yield a proper list.

Maybe it happen implicitly too, but it's better to be explicit.

**Review by spantaleev — CHANGES_REQUESTED** [review]
Looks good! I have a few suggestions for improvements

File: roles/matrix-conduit/templates/conduit/conduit.toml.j2 (hunk 2 above)
Comment (cvwright) [reply roles/matrix-conduit/templates/conduit/conduit.toml.j2:45]: Heh, as you can probably see, I know next to nothing about Jinja templating, and I have only a basic working knowledge of Ansible.  Thanks for the suggestions!

I implemented everything above.  Now running through a full end-to-end test on my system before I commit.

File: roles/matrix-conduit/tasks/setup_conduit.yml [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  when: "matrix_conduit_enabled|bool"
+
+- name: Ensure Conduit data path exists
+  file:
+    path: "{{ matrix_conduit_data_path }}"
+    state: directory
+    mode: 0770
+    owner: "{{ matrix_user_username }}"
+    group: "{{ matrix_user_groupname }}"
+  when: "matrix_conduit_enabled|bool"
```
[the remark this answers is not in the document]
Comment (cvwright) [reply roles/matrix-conduit/tasks/setup_conduit.yml]: I moved the contents of `tasks/setup_conduit.yml` into `tasks/conduit/setup.yml`.

For some reason that I don't quite understand, it doesn't work when I put the directory creation tasks in `tasks/conduit/setup_install.yml`.

[the remark this answers is not in the document]
Comment (spantaleev) [reply roles/matrix-conduit/tasks/setup_conduit.yml]: Hmm, strange. I don't see why it wouldn't work in `tasks/conduit/setup_install.yml`. When putting them in `tasks/conduit/setup_install.yml`, they won't even need the `when` lines.

**Comment by spantaleev** [comment]
Would you add `ansible.builtin.` to these builtin task modules.

Example:

- `set_fact` -> `ansible.builtin.set_fact`
- `template` -> `ansible.builtin.template`

It's probably just for `docker_image` that you can't use it, since it's not part of the `ansible.builtin` collection.

----

`ansible-lint` also asks us to put spaces around filters. Examples:

- `something|to_json` -> `something | to_json`
- `matrix_conduit_enabled|bool` -> `matrix_conduit_enabled | bool`

Could you fix these up as well?

---

Our `ansible-lint` CI check should have told you about these, but I'm not sure why it's not triggering.

**Comment by cvwright** [comment]
> Would you add `ansible.builtin.` to these builtin task modules.
> 
> Example:
> 
> * `set_fact` -> `ansible.builtin.set_fact`
> * `template` -> `ansible.builtin.template`
> 
> It's probably just for `docker_image` that you can't use it, since it's not part of the `ansible.builtin` collection.

Should I change `service` to `ansible.builtin.systemd` as well?

File: roles/matrix-conduit/tasks/setup_conduit.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (cvwright) [reply roles/matrix-conduit/tasks/setup_conduit.yml]: Yeah, I'm definitely doing something wrong here.

I switched my test server back to Synapse by setting `homeserver_implementation` back to "synapse", and the Conduit stuff didn't get removed properly.  So I had to go and manually delete things before Synapse could start working.

I'll take another look on Monday or Tuesday.  I'll start by moving those tasks to `setup_install.yml` as suggested.  Then I'll look into why my Nginx configuration didn't switch over properly.

[the remark this answers is not in the document]
Comment (cvwright) [reply roles/matrix-conduit/tasks/setup_conduit.yml]: Ok, I finally got back to this and moved those lines into `setup_install.yml`.

I think that finishes off the first round of requsted changes.

**Review by spantaleev — APPROVED** [review]

**Comment by spantaleev** [comment]
Thank you for this major feature! :+1: 

It's good to know that the playbook supports 3 different homeserver implementations now!

**Comment by cvwright** [comment]
Very excited to see this land!  Thanks so much for your help!!

**Pull request merged**